File: moboalien-input-suite/java_client/controller_lib/HandshakeClient.java
package java_client.controller_lib;

import java.io.IOException;
import java.net.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Arrays;
import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;

public class HandshakeClient {
    private static final int TIMEOUT_MS = 2000;
    
    public static class HandshakeResult {
        public boolean success;
        public String message;
        public int controllerPort;
        public int screenPort;
        public long serverTicks;
        public PacketEncryptor encryptor;

        public HandshakeResult(boolean success, String message) {
            this.success = success;
            this.message = message;
        }
    }

    public HandshakeResult performHandshake(String serverIP, int handshakePort, String password) {
        try (DatagramSocket socket = new DatagramSocket()) {
            socket.setSoTimeout(TIMEOUT_MS);
            InetAddress serverAddress = InetAddress.getByName(serverIP);
            System.out.println("HandshakeClient: Sending Hello to " + serverAddress + ":" + handshakePort);

            // 1. Send Hello
            byte[] helloPacket = ControllerProtocol.createHelloPacket(1, 0);
            sendPacket(socket, helloPacket, serverAddress, handshakePort);

            // 2. Receive Hello Response
            byte[] buffer = new byte[1024];
            DatagramPacket responsePacket = new DatagramPacket(buffer, buffer.length);
            socket.receive(responsePacket);

            // Parse hello response using ControllerProtocol helper
            byte[] respData = new byte[responsePacket.getLength()];
            System.arraycopy(responsePacket.getData(), 0, respData, 0, responsePacket.getLength());
            ControllerProtocol.HelloResponse helloResp;
            try {
                helloResp = ControllerProtocol.parseHelloResponse(respData);
            } catch (IllegalArgumentException ex) {
                return new HandshakeResult(false, "Invalid hello response: " + ex.getMessage());
            }
            int controllerPort = helloResp.controllerPort;
            int screenPort = helloResp.screenPort;
            byte[] salt = helloResp.salt;

            // Log salt for verification
            System.out.println("Salt (utf8): " + new String(salt, StandardCharsets.UTF_8) + " controllerPort=" + controllerPort + " screenPort=" + screenPort);

            // 3. Send Auth Packet
            byte[] sessionKey = generateSessionKey(password, salt);
            ICryptoHelper cryptoHelper = new StandardCryptoHelper();
            
            // Create challenge payload: "MAGIC" + "Challenge123456"
            String payloadStr = "MAGIC" + "Challenge123456";
            byte[] payload = payloadStr.getBytes(StandardCharsets.UTF_8);
            
            // Encrypt challenge using PacketEncryptor helper (OpenSSL-compatible)
            byte[] encryptedChallenge = cryptoHelper.AESEncrypt(payload, sessionKey);
            System.out.println("Encrypted Challenge (hex): " + bytesToHex(encryptedChallenge));
            
            byte[] authPacket = ControllerProtocol.createAuthPacket(encryptedChallenge);
            sendPacket(socket, authPacket, serverAddress, handshakePort);

            // 4. Receive Auth Result
            socket.receive(responsePacket);
            ByteBuffer respBuffer = ByteBuffer.wrap(responsePacket.getData(), 0, responsePacket.getLength());
            respBuffer.order(ByteOrder.BIG_ENDIAN);

            byte type = respBuffer.get();
            if (type != ControllerProtocol.PACKET_AUTH_RESULT) {
                return new HandshakeResult(false, "Invalid auth result type: " + type);
            }

            byte result = respBuffer.get();
            long serverTicks = respBuffer.getLong();

            if (result != 1) {
                return new HandshakeResult(false, "Authentication failed (Server rejected password)");
            }

            // Handshake successful
            HandshakeResult res = new HandshakeResult(true, "Success");
            res.controllerPort = controllerPort;
            res.screenPort = screenPort;
            res.serverTicks = serverTicks;
            
            res.encryptor = new PacketEncryptor(cryptoHelper, sessionKey, serverTicks);
            return res;

        } catch (Exception e) {
            return new HandshakeResult(false, "Handshake exception: " + e.getMessage());
        }
    }

    private void sendPacket(DatagramSocket socket, byte[] data, InetAddress address, int port) throws IOException {
        DatagramPacket packet = new DatagramPacket(data, data.length, address, port);
        socket.send(packet);
    }

    private byte[] generateSessionKey(String password, byte[] salt) throws Exception {
        MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
        
        // Step 1: Hash the password
        byte[] passHash = sha256.digest(password.getBytes(StandardCharsets.UTF_8));
        String passHashHex = bytesToHex(passHash);
        System.out.println("Password Hash: " + passHashHex);
        
        // Step 2: Decode salt bytes as UTF-8 string
        String saltString = new String(salt, StandardCharsets.UTF_8);
        System.out.println("Salt = " + saltString);
        
        // Step 3: Concatenate password hash (hex string) with salt string, then hash
        // This matches C++: HashText(passwordHash + saltString)
        String combined = passHashHex + saltString;
        byte[] keyHash = sha256.digest(combined.getBytes(StandardCharsets.UTF_8));
        
        // Step 4: Take first 16 bytes for the encryption key
        byte[] finalKey = Arrays.copyOf(keyHash, 16); // Truncate to 16 bytes (128-bit) for compatibility
        System.out.println("Generated Session Key: " + bytesToHex(finalKey));
        return finalKey;
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b & 0xFF));
        }
        return sb.toString();
    }
}

File: moboalien-input-suite/client/handshake_client.cpp
#include "handshake_client.h"
#include "packet_codec.h"
#include "utils.h"
#include "i_crypto_helper.h"
#include <iostream>

namespace Controller {

HandshakeClient::HandshakeClient(Platform* platform)
    : platform_(platform), m_clientTicks(platform->GetTickCountMs()) {
}

HandshakeResult HandshakeClient::PerformHandshake(const std::string& serverIP, int handshakePort, const std::string& password) {
    // Create UDP socket
    uintptr_t sock = platform_->CreateUDPSocket(0);
    if (sock == Platform::INVALID_SOCKET_HANDLE) {
        return HandshakeResult(false, "Socket creation failed");
    }

    // Perform handshake steps
    if (!SendHelloPacket(sock, serverIP, handshakePort)) {
        platform_->CloseSocket(sock);
        return HandshakeResult(false, "Failed to send Hello packet");
    }

    std::vector<uint8_t> salt;
    uint16_t controllerPort, screenPort;
    if (!ReceiveHelloResponse(sock, salt, controllerPort, screenPort)) {
        platform_->CloseSocket(sock);
        return HandshakeResult(false, "Failed to receive Hello response");
    }

    if (!SendAuthPacket(sock, serverIP, handshakePort, password, salt)) {
        platform_->CloseSocket(sock);
        return HandshakeResult(false, "Failed to send Auth packet");
    }

    uint64_t serverTicks;
    if (!ReceiveAuthResult(sock, serverTicks)) {
        platform_->CloseSocket(sock);
        return HandshakeResult(false, "Authentication failed");
    }

    platform_->CloseSocket(sock);
    
    auto cryptoHelper = CreateCryptoHelper();
    // Hash the password first
    std::string passwordHash = cryptoHelper->HashText(password);
    std::cout<<"Password Hash: "<<passwordHash<<std::endl;
    std::cout<<"Salt (hex): "<<BytesToHex(salt)<<std::endl;
    // Concatenate password hash (hex string) bytes with raw salt bytes
    std::vector<uint8_t> passHashBytes(passwordHash.begin(), passwordHash.end());
    std::vector<uint8_t> combined = passHashBytes;
    combined.insert(combined.end(), salt.begin(), salt.end());
    // Hash the combined bytes
    std::string combinedStr(combined.begin(), combined.end());
    std::string keyhash = cryptoHelper->HashText(combinedStr);
    std::vector<uint8_t> encryptionKey = *HexToBytes(keyhash);
    encryptionKey.resize(16); // Use 128-bit key for Java compatibility
    m_packetEncryptor = std::make_shared<PacketEncryptor>(
        cryptoHelper,  // Pass the shared_ptr directly
        platform_,
        serverTicks,
        m_clientTicks,
        encryptionKey
    );
    
    HandshakeResult result(true);
    result.salt = salt;
    result.controllerPort = controllerPort;
    result.screenPort = screenPort;
    result.serverTicks = serverTicks;
    return result;
}

bool HandshakeClient::SendHelloPacket(uintptr_t sock, const std::string& serverIP, int handshakePort) {
    auto helloPacket = CreateHelloPacket(1, 0);
    int sent = platform_->SendTo(sock, helloPacket.data(), helloPacket.size(), 0, serverIP, handshakePort);
    return sent > 0;
}

bool HandshakeClient::ReceiveHelloResponse(uintptr_t sock, std::vector<uint8_t>& outSalt, uint16_t& outControllerPort, uint16_t& outScreenPort) {
    char buffer[1024];
    std::string senderIP;
    int senderPort;
    int len = platform_->RecvFrom(sock, buffer, sizeof(buffer), 0, senderIP, senderPort);
    
    if (len <= 0 || static_cast<uint8_t>(buffer[0]) != PACKET_HELLO_RESPONSE) {
        return false;
    }
    
    HelloResponsePacket helloResp;
    if (!ParseHelloResponsePacket(buffer + 1, len - 1, helloResp, outSalt)) {
        return false;
    }
    
    outControllerPort = helloResp.controllerPort;
    outScreenPort = helloResp.screenPort;
    return true;
}

bool HandshakeClient::SendAuthPacket(uintptr_t sock, const std::string& serverIP, int handshakePort, 
                                     const std::string& password, const std::vector<uint8_t>& salt) {
    auto crypto = CreateCryptoHelper();
    std::string passwordHash = crypto->HashText(password);
    std::string saltStr(salt.begin(), salt.end());
    std::string sessionKeyHash = crypto->HashText(passwordHash + saltStr);
    
    // Create a random challenge
    std::string challenge = "Challenge123456"; // In real app, use random bytes
    std::string payload = "MAGIC" + challenge;
    
    // Encrypt challenge with session key
    auto key = HexToBytes(sessionKeyHash);
    if (!key) {
        return false;
    }
    key->resize(16); // Use 128-bit key for Java compatibility
    
    std::vector<uint8_t> encrypted = std::vector<uint8_t>(payload.begin(), payload.end());
    crypto->AESEncrypt(encrypted, *key);
    auto authPacket = CreateAuthPacket(encrypted);
    
    int sent = platform_->SendTo(sock, authPacket.data(), authPacket.size(), 0, serverIP, handshakePort);
    return sent > 0;
}

bool HandshakeClient::ReceiveAuthResult(uintptr_t sock, uint64_t& outServerTicks) {
    char buffer[1024];
    std::string senderIP;
    int senderPort;
    int len = platform_->RecvFrom(sock, buffer, sizeof(buffer), 0, senderIP, senderPort);
    
    if (len <= 0 || static_cast<uint8_t>(buffer[0]) != PACKET_AUTH_RESULT) {
        return false;
    }
    
    AuthResultPacket authResult;
    if (!ParseAuthResultPacket(buffer + 1, len - 1, authResult)) {
        return false;
    }
    
    outServerTicks = authResult.serverTicks;
    return authResult.result == 1;
}

} // namespace Controller
